### PR TITLE
translation types

### PR DESCRIPTION
**Pull request by ice-orion**
Add script that generates types for translation tags + translation params.
Set this step as part of the `fetch-translation` script + add `fetch-translation` to the precommit-hook.
Add typesafety to the `translate` function + pluralization typing, thanks to @ice-atlas for pointing out 
Group executable scripts into the `scripts` folder.

### CODE REVIEW

File: src/translations/i18n.ts [hunk 1]
```diff
@@ -22,7 +21,14 @@ i18n.translations = locales;
...
 
-export const translate = (scope: string, options?: TranslateOptions) =>
-  i18n.t(scope, options);
+export function translate<
+  T extends keyof Translations,
+  S extends {
+    [key in Translations[T][number]]: string | number;
+  } & TranslateOptions,
+  // @ts-ignore TODO::fix warning
+>(tag: T, options: S = {}) {
```
Comment (ice-atlas) [inline src/translations/i18n.ts:30]: let's keep name as everyone know it. `key` and `options`. So `tag` -> `key`, `T` -> `K`, `S` -> `O`

**Review by ice-atlas — CHANGES_REQUESTED** [review]

File: src/translations/i18n.ts [hunk 2]
```diff
@@ -22,7 +21,14 @@ i18n.translations = locales;
...
 export default i18n;
 
-export const translate = (scope: string, options?: TranslateOptions) =>
-  i18n.t(scope, options);
+export function translate<
+  T extends keyof Translations,
+  S extends {
+    [key in Translations[T][number]]: string | number;
+  } & TranslateOptions,
+  // @ts-ignore TODO::fix warning
```
Comment (ice-ares) [inline src/translations/i18n.ts:29]: fix it

File: scripts/generateTranslationTypes/utils/getParamsFromString.mjs [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export const getParamsFromString = input => {
+  return [...input.matchAll(/{{(.*?)}}/g)].map(match => match[1]);
```
Comment (ice-ares) [inline scripts/generateTranslationTypes/utils/getParamsFromString.mjs:2]: be more strict, allow atleast 1 letter, and only alphanumerics

**Review by ice-ares — CHANGES_REQUESTED** [review]
revert the license.header change. That file has to stay at project root

**Comment by ice-atlas** [comment]
@ice-ares 
> revert the license.header change. That file has to stay at project root

Maybe better use `LICENSE` file for this and leave `license.header` as is now?

**Comment by ice-ares** [comment]
no, both license and license header have to be at project root

> @ice-ares 
> 
> > revert the license.header change. That file has to stay at project root
> 
> 
> 
> Maybe better use `LICENSE` file for this and leave `license.header` as is now?

File: package.json [hunk 4]
```diff
@@ -13,11 +13,12 @@
...
     "match:renew": "bundler exec fastlane match adhoc --force_for_new_devices",
     "test": "jest tests --coverage --updateSnapshot --passWithNoTests",
-    "fetch-translations": "source .env && curl \"https://localise.biz/api/export/all.json?format=multi&status=translated&fallback=en&key=$LOCALIZE_BIZ_TOKEN\" --output src/translations/translations.json",
+    "fetch-translations": "source .env && curl \"https://localise.biz/api/export/all.json?format=multi&status=translated&fallback=en&key=$LOCALIZE_BIZ_TOKEN\" --output src/translations/translations.json && yarn generate-translation-types",
+    "generate-translation-types": "scripts/generateTranslationTypes/index.mjs src/translations/translations.json && eslint --fix src/translations/translations.json.d.ts",
     "lint": "eslint . --ext .js,.jsx,.ts,.tsx",
     "check-types": "tsc",
     "postinstall": "npx jetify && npx patch-package",
-    "check-license": "./checklicense.sh",
+    "check-license": "./scripts/checkLicense/checklicense.sh",
```
[the remark this answers is not in the document]
Comment (ice-orion) [reply package.json]: done

File: scripts/run-e2e-android-release.sh [hunk 5]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 source .env
-yarn start &
```
[the remark this answers is not in the document]
Comment (ice-orion) [reply scripts/run-e2e-android-release.sh]: done

File: scripts/run-e2e-ios-release.sh [hunk 6]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 source .env
-yarn start &
```
[the remark this answers is not in the document]
Comment (ice-orion) [reply scripts/run-e2e-ios-release.sh]: done

File: src/screens/InviteFlow/InviteShare/components/ShareButton/index.tsx [hunk 7]
```diff
@@ -23,7 +24,7 @@ export type SocialType =
 
 export type SocialShareButtonType = {
   type: SocialType;
-  title: string;
+  title: keyof Translations;
```
[the remark this answers is not in the document]
Comment (ice-orion) [reply src/screens/InviteFlow/InviteShare/components/ShareButton/index.tsx]: done

File: src/translations/i18n.ts (hunk 1 above)
Comment (ice-orion) [reply src/translations/i18n.ts]: done

File: scripts/generateTranslationTypes/utils/getParamsFromString.mjs (hunk 3 above)
Comment (ice-orion) [reply scripts/generateTranslationTypes/utils/getParamsFromString.mjs]: done

File: src/translations/i18n.ts (hunk 2 above)
Comment (ice-orion) [reply src/translations/i18n.ts]: done

**Review by ice-ares — APPROVED** [review]

**Review by ice-ares — APPROVED** [review]

File: src/services/auth/signin/apple.ts [hunk 8]
```diff
@@ -43,7 +43,7 @@ export const startAppleSignIn: SocialSignInMethod<{
           return {cancelled: true};
         default:
           throw new Error(
-            t('errors.unknown_error_with_code', {code: error.code}),
+            t('errors.unknown_error_with_code', {code: error.code as string}),
```
Comment (ice-hector) [inline src/services/auth/signin/apple.ts:46]: Can we avoid force type casting here?

File: src/services/auth/signin/google.ts [hunk 9]
```diff
@@ -48,7 +48,7 @@ export const startGoogleSignIn: SocialSignInMethod<{
           throw new Error(t('errors.google_play_services_not_available'));
         default:
           throw new Error(
-            t('errors.unknown_error_with_code', {code: error.code}),
+            t('errors.unknown_error_with_code', {code: error.code as string}),
```
Comment (ice-hector) [inline src/services/auth/signin/google.ts:51]: Same here

File: src/services/auth/signin/apple.ts (hunk 8 above)
Comment (ice-orion) [reply src/services/auth/signin/apple.ts]: Yea, we actually can convert it to string, fixed

File: src/services/auth/signin/google.ts (hunk 9 above)
Comment (ice-orion) [reply src/services/auth/signin/google.ts]: fixed

**Review by ice-hector — APPROVED** [review]

**Pull request merged**